src/test/coll.rs

```diff
@@ -1201,3 +1201,216 @@ fn assert_duplicate_key_error_with_utf8_replacement(error: &ErrorKind) {
+#[cfg_attr(feature = "async-std-runtime", async_std::test)]
+async fn bulk_write_failure_has_inserted_ids() {
+    let _guard: RwLockReadGuard<()> = LOCK.run_concurrently().await;
+
+    let client = TestClient::new().await;
+
+    // an ordered, single batch bulk write where the last doc generates a write error:
+    // everything before the last doc should be inserted
+    let docs = vec![doc! { "_id": 2}, doc! { "_id": 1}];
+    let res = run_inserted_ids_test(&client, &docs, true);
+    let err = res.await.expect_err("insert_many should fail");
+    match *err.kind {
+        ErrorKind::BulkWrite(failure) => {
+            assert_eq!(
+                failure.inserted_ids.len(),
+                1,
+                "one document should have been inserted"
+            );
+            assert!(
+                failure.inserted_ids.contains_key(&0),
+                "document at index 0 should have been inserted"
+            );
+            assert_eq!(
+                failure.inserted_ids.get(&0).unwrap(),
+                &Bson::Int32(2),
+                "inserted document should have _id 2"
+            );
+        }
+        _ => panic!("Expected BulkWrite error, but got: {:?}", err),
+    }
+
+    // an ordered, single batch bulk write where the first doc generates a write error:
+    // nothing should be inserted
+    let docs = vec![doc! { "_id": 1}, doc! { "_id": 2}];
+    let res = run_inserted_ids_test(&client, &docs, true);
+    let err = res.await.expect_err("insert_many should fail");
+    match *err.kind {
+        ErrorKind::BulkWrite(failure) => {
+            assert_eq!(
+                failure.inserted_ids.len(),
+                0,
+                "inserted_ids should be empty"
+            );
+        }
+        _ => panic!("Expected BulkWrite error, but got: {:?}", err),
+    }
+
+    // an unordered, single batch bulk write where the first doc generates a write error:
+    // everything after the first doc should be inserted
+    let res = run_inserted_ids_test(&client, &docs, false);
+    let err = res.await.expect_err("insert_many should fail");
+    match *err.kind {
+        ErrorKind::BulkWrite(failure) => {
+            assert_eq!(
+                failure.inserted_ids.len(),
+                1,
+                "one document should have been inserted"
+            );
+            assert!(
+                failure.inserted_ids.contains_key(&1),
+                "document at index 1 should have been inserted"
+            );
+            assert_eq!(
+                failure.inserted_ids.get(&1).unwrap(),
+                &Bson::Int32(2),
+                "inserted document should have _id 2"
+            );
+        }
+        _ => panic!("Expected BulkWrite error, but got: {:?}", err),
+    }
+
+    // an ordered, 2-batch bulk write where a document in the first batch generates write error:
+    // nothing should be inserted
+    // note: these numbers were chosen because maxWriteBatchSize is 100,000
```

Comment:
@isabelatkinson for the purpose of trying to write spec tests around batching behavior for bulk, we might consider requiring drivers have some way to internally override `maxWriteBatchSize`. otherwise we'd need  massive yaml/json files to fit in enough data to generate multiple batches (the limit is either 100k writes or 16 MB, whichever is smaller). 
alternatively, drivers can write prose tests like this one, but I think it would be much easier if we could artificially use bulk batches with a smaller size.
alternatively/additionally we could consider asking the server to make the parameter somehow configurable so we could temporarily decrease it for the bulk tests